### PR TITLE
[feature/semantic_text] Refactor inference to run as an action filter

### PR DESCRIPTION
**Pull request by jimczi**
This PR moves the implementation of the ingest bulk inference back to the inference plugin. 
With all the ML dependencies it is challenging to rely on things available in server and every new feature (chunking, bulk inference, ...) requires to move things around without clear boundaries.
The action filter is a good compromise at the moment since it allows to modify the bulk shard request transparently and to ensure that we have all the code for semantic_text in the same place.
This draft PR handles chunking and executes all the inference per service in a bulk fashion.

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * An {@link ActionFilter} that performs inference on {@link BulkShardRequest} asynchronously and stores the results in
+ * the individual {@link BulkItemRequest}. The results are then consumed by the {@link SemanticTextInferenceResultFieldMapper}
+ * in the subsequent {@link TransportShardBulkAction} downstream.
+ */
+public class ShardBulkInferenceActionFilter implements ActionFilter {
```
Comment (jimczi) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java:56]: That's where the ingest inference entire logic is implemented.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            }
+
+            final IndexRequest indexRequest = getIndexRequestOrNull(item.request());
+            Map<String, Object> newDocMap = indexRequest.sourceAsMap();
+            Map<String, Object> inferenceMap = new LinkedHashMap<>();
+            // ignore the existing inference map if any
+            newDocMap.put(InferenceResultFieldMapper.NAME, inferenceMap);
+            for (FieldInferenceResponse fieldResponse : response.responses()) {
+                try {
+                    InferenceResultFieldMapper.applyFieldInference(
```
Comment (carlosdelest) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java:267]: Nice - now we can use common code for this

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    @Override
+    public <Request extends ActionRequest, Response extends ActionResponse> void apply(
+        Task task,
+        String action,
+        Request request,
+        ActionListener<Response> listener,
+        ActionFilterChain<Request, Response> chain
+    ) {
+        switch (action) {
+            case TransportShardBulkAction.ACTION_NAME:
```
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java:81]: No, we decorate the shard bulk request [here](https://github.com/elastic/elasticsearch/blob/86ddc9d8b8aff0285890732d2164c589b87ce3dc/server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java#L213) and then calls the action [locally](https://github.com/elastic/elasticsearch/blob/86ddc9d8b8aff0285890732d2164c589b87ce3dc/server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java#L227). So the filter will only intercepts the transport shard bulk action that is local to the parent bulk action.

[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java:81]: Got it, thanks!

**Review by carlosdelest — APPROVED** [review]
LGTM, amazing refactoring!

**Comment by gwbrown** [comment]
Apologies for the label noise/ping, did stuff in the wrong tab while triaging `needs:triage` issues.

**Review by rjernst — COMMENTED** [review]
I haven't looked through this entire PR since it is quite large. In general it looks like it removes a lot of abstractions from server, so that seems good!

There's one important change though I think is needed. Action filters have a cost. They are executed on _every_ transport action, regardless of whether it's the action you actually want to filter. However, there is a newer mechanism for this: MappedActionFilter. If you implement that, it won't have any new runtime cost to other transport actions.

Second

**Comment by jimczi** [comment]
Thanks for looking @rjernst . Note that this PR is targeting a branch and not main since we're still iterating on the feature. Although I am glad you checked and I applied your suggestion.

**Pull request merged**